### PR TITLE
Add new listing to inventory/marketplace

### PR DESCRIPTION
**Pull request by AnssiAhola**
Implements [this](https://www.discogs.com/developers#page:marketplace,header:marketplace-new-listing) API feature.
Also fixes issues described in this discussion https://github.com/joalla/discogs_client/discussions/17#discussion-1930401

Needs testing for potential issues or side effects.

### CODE REVIEW

**Review by JOJ0 — CHANGES_REQUESTED** [review]
Hi @AnssiAhola thank you very much for this nice new feature! :-), and certainly also for fixing these bugs in client anf fetchers! Greatly appreciated!

There are some trailing spaces in lines 402/403 in models.py. Please remove them.

Also it would be good to add a test case for this class/method. Have a look at the Wantlist and Collection tests. They might help you with how it should roughly look like: 

https://github.com/joalla/discogs_client/blob/master/discogs_client/tests/test_models.py#L123
https://github.com/joalla/discogs_client/blob/master/discogs_client/tests/test_models.py#L148

**Comment by AnssiAhola** [comment]
Adde the test.
What's the reason for mocking the expected responses in the tests?

**Comment by JOJ0** [comment]
> Added the test.
Thanks for the test! Looks alright to me. 

> What's the reason for mocking the expected responses in the tests?
That is a very good question indeed....and I can't tell you by heart even though I did write it back in June 2020.

@hellricer helped me back then and I think this is the answer:
https://github.com/joalla/discogs_client/pull/3#discussion_r438664911

I am not sure if you need this mocking at all in your case. If the test works, leave it out!

**Review by JOJ0 — APPROVED** [review]

**Comment by JOJ0** [comment]
@alifhughes if you find the time please give this PR your final review and press the button! :-)

**Comment by alifhughes** [comment]
Sorry I haven't forgotten this, I plan to review this in the next 24 hours. Sorry for the delay!

File: discogs_client/models.py [hunk 1]
```diff
@@ -397,6 +397,29 @@ def remove(self, release):
         self.client._delete(self.url + '/' + str(release_id))
         self._invalidate()
 
+
+class Inventory(PaginatedList):
+    def add_listing(self, release, condition, price, status, sleeve_condition=None,
```
Comment (alifhughes) [inline discogs_client/models.py:402]: I see that we already have a `Listing` class down at the bottom of this file - [here](https://github.com/joalla/discogs_client/blob/master/discogs_client/models.py#L642-L662) - which interacts with the same endpoint `/marketplace/listings`. I see that the `Listings` class has _some_ of the data you want to post, maybe you could alter that class to do what you want? What do you think?

Comment (alifhughes) [inline discogs_client/models.py:402]: If you **_do_** merge this class with `Listings`, I think it would make sense to rename this function to `add()` instead, as that is the convention with other `add()` functions in these models

Comment (AnssiAhola) [reply discogs_client/models.py:402]: I considered it, but that `Listing` class is for listings that already exist in the inventory.
`Listing` class requires `listing_id` on init as its included in the resource_url.
But if there's a better solution I'm missing, let me know and I'll fix it 🙂

**Comment by AnssiAhola** [comment]
Is it ok if I add [these ](https://www.discogs.com/developers#page:marketplace,header:marketplace-listing)"seller only" fields to the listing in this PR?
They will return `None` if fetched as non-seller.

**Comment by alifhughes** [comment]
> Is it ok if I add [these ](https://www.discogs.com/developers#page:marketplace,header:marketplace-listing)"seller only" fields to the listing in this PR?
> They will return `None` if fetched as non-seller.
> 
> Or should I create a new PR as I'm thinking i should add update/change method to Listing.

@AnssiAhola I think if possible we could separate into a different PR, especially if you want to change a different method that is unrelated to the original idea of this PR. I'm always keen to keep PRs small and focused if it makes sense to!

File: discogs_client/client.py [hunk 2]
```diff
@@ -95,7 +95,8 @@ def _request(self, method, url, data=None):
         }
 
         if data:
-            headers['Content-Type'] = 'application/x-www-form-urlencoded'
+            headers['Content-Type'] = 'application/json'
```
Comment (alifhughes) [inline discogs_client/client.py:98]: I think we have to be a little careful with this change, these `headers` are used in the `OAuth2Fetcher` class's `self.client.sign()` function. From [the docs](https://oauthlib.readthedocs.io/en/latest/oauth1/client.html#) it seems to state that we can only use `x-www-url-formencoded`:
> **Request body**
>     The OAuth 1 spec only covers signing of x-www-url-formencoded information.

I had a play about with this change only (not the `data`-> `body` change as well which could be the issue, but see below about that) the change and I seem to now get an error with the `sign()` function:

```
>>> import discogs_client
>>> d = discogs_client.Client('ExampleApplication/0.1')
>>> d.set_consumer_key('key-here', 'secret-here')
>>> d.get_authorize_url()
>>> d.get_access_token('verifier-here')
>>> me = d.identity()
>>> me.wantlist.add(d.release(5))

(added breakpoint in first line of `OAuth2Fetcher.fetch())

>>> data
{'release_id': '5'}
>>> headers
{'Accept-Encoding': 'gzip', 'User-Agent': 'ExampleApplication/0.1', 'Content-Type': 'application/json'}

(continued)

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/alistair.hughes/Documents/joalla/discogs_client/discogs_client/models.py", line 392, in add
    self.client._put(self.url + '/' + str(release_id), omit_none(data))
  File "/Users/alistair.hughes/Documents/joalla/discogs_client/discogs_client/client.py", line 125, in _put
    return self._request('PUT', url, data)
  File "/Users/alistair.hughes/Documents/joalla/discogs_client/discogs_client/client.py", line 100, in _request
    content, status_code = self._fetcher.fetch(self, method, url, data=data, headers=headers)
  File "/Users/alistair.hughes/Documents/joalla/discogs_client/discogs_client/fetchers.py", line 98, in fetch
    body = json.dumps(data) if json_format and data else data
  File "/Users/alistair.hughes/Documents/joalla/discogs_client/env/lib/python3.7/site-packages/oauthlib/oauth1/rfc5849/__init__.py", line 274, in sign
    "Body contains parameters but Content-Type header was not set.")
ValueError: Body contains parameters but Content-Type header was not set.
```

Not sure if I'm missing something? Have you tried this flow? Again from below, **we might have to start passing around `json_format=False` to get these things to all work harmoniously.** Let me know if I've not tested this correctly!

File: discogs_client/fetchers.py [hunk 3]
```diff
@@ -96,7 +96,7 @@ def set_verifier(self, verifier):
     def fetch(self, client, method, url, data=None, headers=None, json_format=True):
         body = json.dumps(data) if json_format and data else data
         uri, headers, body = self.client.sign(url, http_method=method,
-                                              body=data, headers=headers)
+                                              body=body, headers=headers)
```
Comment (alifhughes) [inline discogs_client/fetchers.py:99]: Similar to the comment above, I seem to be getting errors now with this change with this which is related to the headers change also, do the following steps:

```python
>>> import discogs_client
>>> d = discogs_client.Client('ExampleApplication/0.1')
>>> d.set_consumer_key('key-here', 'secret-here')
>>> d.get_authorize_url()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/alistair.hughes/Documents/joalla/discogs_client/discogs_client/client.py", line 53, in get_authorize_url
    content, status_code = self._fetcher.fetch(self, 'POST', self._request_token_url, data=postdata, headers=params)
  File "/Users/alistair.hughes/Documents/joalla/discogs_client/discogs_client/fetchers.py", line 98, in fetch
    body = json.dumps(data) if json_format and data else data
  File "/Users/alistair.hughes/Documents/joalla/discogs_client/env/lib/python3.7/site-packages/o authlib/oauth1/rfc5849/__init__.py", line 269, 
in sign
    "Headers indicate a formencoded body but body was not decodable.")
ValueError: Headers indicate a formencoded body but body was not decodable.
```

I added a breakpoint and can see that:
```
>>> headers = {'User-Agent': 'ExampleApplication/0.1', 'Content-Type': 'application/x-www-form-urlencoded'}
>>> data = 'User-Agent=ExampleApplication%2F0.1&Content-Type=application%2Fx-www-form-urlencoded'
>>> body = '"User-Agent=ExampleApplication%2F0.1&Content-Type=application%2Fx-www-form-urlencoded"'
```
If I understand it correctly, with this change we're forcing the `json_dumps` but in the case of `get_authorize_url()` it does not need to be and causes the error? I think we need to consider using the `json_format` param to set to false in these cases. Do you get a similar error?

**Review by alifhughes — CHANGES_REQUESTED** [review]
Left a couple of comments, please let me know if I'm not getting the right picture/not testing correctly!

**Comment by AnssiAhola** [comment]
get_authorize_url() defines it's own headers so it's not using the `application/json` one, its just the `json_format` flag that's the issue I think.
https://github.com/joalla/discogs_client/blob/f4d6efcc89950c41ceb37f4b718430dcdac9aea8/discogs_client/client.py#L48-L54

**Comment by AnssiAhola** [comment]
that headers thing is a bit strange, that oauthlib doc you linked sure says `x-www-url-formencoded`, but discogs requires `/json` when POSTing data or else it'll give response 400: Not a JSON request 🤔. If I've understood correctly, that is 😄

File: discogs_client/fetchers.py (hunk 3 above)
Comment (alifhughes) [reply discogs_client/fetchers.py:99]: > I've no idea why but I added json_format=False to line 53 (get_authorize_url() in client.py) in #23.
Those two above changes that you tested seperately, need to be used with the line 53 change from #23 for it to function correctly I think, at least on my end.

Ahh this makes sense now 😄 Thanks for explaining. I've tried with the latest from master and your changes and as far as I can test, it seems to work so good job!

> Sorry about that 😰
No worries! It happens and we (as reviewers) didn't catch it either 🙂

File: discogs_client/client.py (hunk 2 above)
Comment (alifhughes) [reply discogs_client/client.py:98]: Answered [here](https://github.com/joalla/discogs_client/pull/18#discussion_r564825880).

**Review by alifhughes — CHANGES_REQUESTED** [review]

**Comment by alifhughes** [comment]
> that headers thing is a bit strange, that oauthlib doc you linked sure says only `x-www-url-formencoded`, but discogs requires `/json` when POSTing data or else it'll give response 400: Not a JSON request 🤔. If I've understood correctly, that is 😄

@AnssiAhola So I've been doing some light digging, this is what I've found:

**TL;DR** - If I understand it correctly, `oauthlib.oauth1.sign()` thinks that there is no body, and therefore it doesn't matter that the `Content-Type` is not `application/x-www-form-urlencoded`. **Whether this is an issue or not**, I'm still unsure 🤔. The changes work as far as I can see, my only concern is that it _**could**_ mean that the API will get 'unsigned' requests, but I would have thought that this would be kicked back in the response?

**Longer version:**
It doesn't recognise there is a body because it is a string as a result of `json_dumps`, [it cannot decode a `str`](https://github.com/oauthlib/oauthlib/blob/d54965b86ce4ede956db70baff0b3d5e9182a007/oauthlib/common.py#L142-L146). Therefore `request.decoded_body = None` [(here)](https://github.com/oauthlib/oauthlib/blob/d54965b86ce4ede956db70baff0b3d5e9182a007/oauthlib/oauth1/rfc5849/__init__.py#L301) this means that `has_params` is `False` and `should_have_params` is also `False` because `content_type != CONTENT_TYPE_FORM_URLENCODED` [(here)](url). [There is no `if` catch if they're both `False`](https://github.com/oauthlib/oauthlib/blob/d54965b86ce4ede956db70baff0b3d5e9182a007/oauthlib/oauth1/rfc5849/__init__.py#L306-L321) because that is fine - why _should_ it have params if it doesn't _actually_ have them?

However, I'm not certain if this is expected or not, I think I'll open an issue to ask this, unless either of you @AnssiAhola or @JOJ0 would happen to know? I think if I want to find the definitive answer, I might have to read the [spec](https://tools.ietf.org/html/rfc5849) but I will do that another time. This is as much as I can work out for the time being and I don't think it can block this PR.

File: discogs_client/client.py [hunk 4]
```diff
@@ -95,7 +95,8 @@ def _request(self, method, url, data=None):
...
+            # headers['Content-Type'] = 'application/x-www-form-urlencoded'
```
[the remark this answers is not in the document]
Comment (AnssiAhola) [reply discogs_client/client.py]: done!

**Comment by AnssiAhola** [comment]
Thanks for the explanation @alifhughes , I've no idea if its expected or not, haha.
But like you said, you'd think the API would give a response if the requests we're unsigned.

File: discogs_client/models.py (hunk 1 above)
Comment (AnssiAhola) [reply discogs_client/models.py:402]: Should I change this from `add_listing()` to `add()`

Comment (alifhughes) [reply discogs_client/models.py:402]: `add_listing()` is fine actually I noticed there are other similar names in the file like `add_release()` [here](https://github.com/joalla/discogs_client/blob/67a4dc9a889d69fc58c9299089de6f1f134fd4a6/discogs_client/models.py#L638)

**Review by alifhughes — APPROVED** [review]
LGTM, thanks for addressing the comments, we'll get a release done ASAP. 🚀

**Comment by alifhughes** [comment]
@AnssiAhola this has been released inversion 2.3.7 - https://github.com/joalla/discogs_client/releases/tag/v2.3.7
